cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -603,9 +603,9 @@ struct BooleanAllImpl : public ScalarAggregator {
     }
     if (batch[0].is_scalar()) {
       const Scalar& scalar = *batch[0].scalar;
-      this->has_nulls = !scalar.is_valid;
-      this->count += scalar.is_valid;
-      this->all = !scalar.is_valid || checked_cast<const BooleanScalar&>(scalar).value;
+      this->has_nulls |= !scalar.is_valid;
+      this->count += scalar.is_valid * batch.length;
+      this->all &= !scalar.is_valid || checked_cast<const BooleanScalar&>(scalar).value;
       return Status::OK();
     }
     const ArraySpan& data = batch[0].array;
```

Comment:
Why didn't change `has_null =` to `has_null |=` as what you did for `any`?

Reply:
done

Reply:
And my question is why make it `|=` instead of simply `=`?

Reply:
The reason is in https://github.com/apache/arrow/pull/43799#issuecomment-2308439586